packages/cli/src/telemetry/index.js

```diff
@@ -0,0 +1,81 @@
+import { spawn } from 'child_process'
+import os from 'os'
+import path from 'path'
+
+import { diag, DiagConsoleLogger, DiagLogLevel } from '@opentelemetry/api'
+import opentelemetry from '@opentelemetry/api'
+import {
+  NodeTracerProvider,
+  BatchSpanProcessor,
+} from '@opentelemetry/sdk-trace-node'
+
+import { CustomFileExporter } from './exporter'
+
+/**
+ * @type NodeTracerProvider
+ */
+let traceProvider
+
+/**
+ * @type BatchSpanProcessor
+ */
+let traceProcessor
+
+/**
+ * @type CustomFileExporter
+ */
+let traceExporter
+
+export async function startTelemetry() {
+  diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.ERROR)
+
+  // Tracing
+  traceProvider = new NodeTracerProvider()
+  traceExporter = new CustomFileExporter()
+  traceProcessor = new BatchSpanProcessor(traceExporter)
+  traceProvider.addSpanProcessor(traceProcessor)
+  traceProvider.register()
+
+  process.on('SIGTERM', async () => {
+    await shutdownTelemetry()
+  })
+  process.on('SIGINT', async () => {
+    // TODO: Should we record a SIGINT as a telemetry event?
+    await shutdownTelemetry()
+  })
+}
+
+export async function shutdownTelemetry() {
+  try {
+    opentelemetry.trace.getActiveSpan()?.end()
+    await traceProvider?.shutdown()
+    await traceProcessor?.shutdown()
+    traceExporter?.shutdown()
+
+    // Send the telemetry in a background process, so we don't block the CLI
+    // We must account for some platform specific behaviour when spawning the process
+    const spawnOptions =
+      os.type() === 'Windows_NT'
+        ? {
+            // The following options run the process in the background without a console window, even though they don't look like they would.
+            // See https://github.com/nodejs/node/issues/21825#issuecomment-503766781 for information
+            detached: false,
+            windowsHide: false,
+            shell: true,
+            stdio: 'inherit',
+          }
+        : {
+            detached: true,
+            stdio: 'inherit',
```

Comment:
If this a detached background process do we want `stdio` to be inherit? I didn't look at send.js at all so maybe something's handled there, or maybe what it's inheriting isn't hooked up to stdio anyway and it doesn't matter?

Reply:
Likely not. I copied what we had for the background update checker since the combination of execa options are a bit magic to get what you really want. There should be no console output but I guess keeping it as inherit might make it obvious if something blows up and starts writing error messages.

Reply:
Yeah but the error would come after the command is over and out of nowhere; maybe we should copy updateChecker exactly and write to a file? I made the send.js file throw just as an example:


https://github.com/redwoodjs/redwood/assets/32992335/686c0bc3-c164-40f7-801b-73391c21d80d